pad-tls/pad-tls.c

```diff
@@ -0,0 +1,112 @@
+/* pad-tls: tool to pad TLS segment of ELF files so that a given compartment's
+TLS variables will only possibly share a page with another compartment's
+padding, not its TLS variables */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <libgen.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+
+#include <elf.h>
+
+enum patch_result {
+  PATCH_SUCCESS,
+  BAD_MAGIC,
+  WRONG_ARCH,
+  NO_TLS,
+};
+
+static const char *patch_errors[] = {
+    "invalid ELF magic number",
+    "ELF architecture is not x86-64",
+    "no TLS segment in ELF program headers",
+};
+
+enum patch_result patch_tls(void *elf, int64_t delta, uint64_t *size_out) {
+  Elf64_Ehdr *ehdr = elf;
+
+  if (memcmp(ehdr->e_ident,
+             "\x7f"
+             "ELF",
+             4)) {
+    return BAD_MAGIC;
+  }
+
+  if (ehdr->e_machine != EM_X86_64) {
+    return WRONG_ARCH;
+  }
+
+  char *ph_addr = ((char *)elf) + ehdr->e_phoff;
+  for (int i = 0; i < ehdr->e_phnum; i++) {
+    Elf64_Phdr *phdr = (Elf64_Phdr *)ph_addr;
+
+    if (phdr->p_type == PT_TLS) {
+      phdr->p_memsz += delta;
```

Comment:
If this is always 4096, do you need it to be a function argument?

Reply:
It's somewhat likely that we'll need to change our padding scheme, possibly on a library-dependent basis, so I figured I should err toward making it easier to make this a command-line parameter eventually, just in case.